### DISCUSSION TITLE
dynamic tensor mechanics axisymmetric RZ

### DISCUSSION THREAD

**Discussion by Alirezarei** [idaholab/moose#21893] [Q&A Modules: Solid mechanics] [answered]
Hello,

I have been trying to use the Dynamic Tensor Mechanics action to solve an axisymmetric RZ problem, but I always get an error saying the coordinate system is not supported. I have tried the following approaches to solve the issue, but nothing has worked so far.

1. I made two ADDynamicStressDivergenceTensors in different directions, but I get an error saying: "The coordinate system in the problem block must be set to XYZ for cartesian geometries"
2. I used Tensor mechanics master action and added two ADInertialForce kernels in different directions but the solve does not converge in this case!
3. I used the Dynamic Master action without specifying the coord_type and set the cylindrical_axis_point1 and cylindrical_axis_point2 to proper vectors but it still solves in cartesian coordinate system!

Could you help me with this please?

I appreciate your time.
Regards,
Alireza

**Comment by hugary1995** [comment]
You are probably missing the following block in the input file
```python
[Mesh]
  coord_type = RZ
[]
```

**Reply by Alirezarei** [reply]
Thanks for your reply. When I add coord_type = RZ and I'm using dynamic master action, it throws an error saying: "unsupported coordinate system". The problem is that Dynamic Master action seems not to support axisymmetric RZ problems or I'm missing something!

**Reply by recuero** [reply]
Yeah, we don't have RZ in dynamics. I recommend you use the following input: https://github.com/idaholab/moose/blob/436a787585e17c3901eb7de29dbd43a80d9a881a/modules/tensor_mechanics/test/tests/dynamics/rayleigh_damping/rayleigh_hht.i where inertia is already included in the latest action. 

You mention the parameters `cylindrical_axis_point1` and `cylindrical_axis_point2`: Aren't those gap heat transfer parameters? In any case, if you want to use dynamics, unfortunately for now, you'll have to stick to Cartesian systems.

**Reply by Alirezarei** [reply]
I see. Thanks for your help.
I really need to perform a dynamic analysis in RZ coordinate system. Are you planning to add this feature to Tensor Mechanics Dynamic Master action any time soon? If not, how can I do it myself?

Also, I'm using Tensor Mechanics Master action that supports RZ coordinate system but the linear solver does not converge due to DIVERGED_PC_FAILED; PC failed due to FACTOR_NUMERIC_ZEROPIVOT. The same input works for XYZ coordinate system. Is there any way that I can solve this issue?

**Comment by recuero** [comment, marked as answer]
> Are you planning to add this feature to Tensor Mechanics Dynamic Master action any time soon? If not, how can I do it myself?

I don't believe we can work on that soon. If you are interested in looking at it, I'd review `DynamicStressDivergenceTensors` which does the Cartesian definition of stresses within a dynamic integrator and then also at `StressDivergenceRZTensors` which does the axisymmetric work for quasi-statics. To successfully implement the RZ problem in dynamics you'd need to derive the expressions including `_alpha`, `_dt`, and `_zeta`. 

>  I'm using Tensor Mechanics Master action that supports RZ coordinate system but the linear solver does not converge due to DIVERGED_PC_FAILED; PC failed due to FACTOR_NUMERIC_ZEROPIVOT

It's hard to know what's going on there. I recommend you look at quasistatic RZ problems in the repository and see if you are missing something. Wonder if your BCs are correct since you mentioned you used the same input for XYZ.

**Reply by Alirezarei** [reply]
Thank you very much for your prompt responses. I really appreciate it. I will check them out and see if I can do it myself.

**Comment by lynnmunday** [comment]
A work around might be to use TotalLagrangianStressDivergenceAxisymmetricCylindrical like in this example:
https://github.com/idaholab/marlin/blob/devel/examples/impact/rz_elastic_nodal.i
